src/chain_sync/tipset_syncer.rs

```diff
@@ -807,111 +811,114 @@ async fn sync_headers_in_reverse<DB: Blockstore + Sync + Send + 'static>(
             .await
             .map_err(TipsetRangeSyncerError::NetworkTipsetQueryFailed)?;
 
-        for tipset in network_tipsets {
-            // Break if have already traversed the entire tipset range
-            if tipset.epoch() < current_head.epoch() {
-                break 'sync;
-            }
-            validate_tipset_against_cache(bad_block_cache, tipset.key(), &parent_blocks)?;
-            parent_blocks.extend(tipset.cids());
+        for tipset in network_tipsets
+            .into_iter()
+            .take_while(|ts| ts.epoch() >= until_epoch)
+        {
+            validate_tipset_against_cache(bad_block_cache, tipset.key(), &accepted_blocks)?;
+            accepted_blocks.extend(tipset.cids());
             tracker.write().set_epoch(tipset.epoch());
-            parent_tipsets.push(tipset);
+            pending_tipsets.push(tipset);
         }
     }
     drop(wp);
 
-    let oldest_tipset = parent_tipsets.last().clone();
-    // Determine if the local chain was a fork.
-    // If it was, then sync the fork tipset range by iteratively walking back
+    let oldest_pending_tipset = pending_tipsets.last();
+    // common case: receiving a block that's potentially part of the same tipset as our best block
+    if oldest_pending_tipset.as_ref() == current_head
+        || oldest_pending_tipset.is_child_of(current_head)
+    {
+        return Ok(pending_tipsets);
+    }
+
+    // Fork detected, sync the fork tipset range by iteratively walking back
     // from the oldest tipset synced until we find a common ancestor
-    if oldest_tipset.parents() != current_head.parents() {
-        info!("Fork detected, searching for a common ancestor between the local chain and the network chain");
-        const FORK_LENGTH_THRESHOLD: u64 = 500;
-        let fork_tipsets = network
-            .chain_exchange_headers(None, oldest_tipset.parents(), FORK_LENGTH_THRESHOLD)
-            .await
-            .map_err(TipsetRangeSyncerError::NetworkTipsetQueryFailed)?;
-        let mut potential_common_ancestor = chain_store
-            .chain_index
-            .load_required_tipset(current_head.parents())?;
-        let mut i = 0;
-        let mut fork_length = 1;
-        while let Some(fork_tipset) = fork_tipsets.get(i) {
-            if fork_tipset.epoch() == 0 {
-                return Err(TipsetRangeSyncerError::ForkAtGenesisBlock(format!(
-                    "{:?}",
-                    oldest_tipset.cids()
-                )));
-            }
-            if &potential_common_ancestor == fork_tipset {
-                // Remove elements from the vector since the Drain
-                // iterator is immediately dropped
-                let mut fork_tipsets = fork_tipsets;
-                fork_tipsets.drain((i + 1)..);
-                parent_tipsets.extend(fork_tipsets);
-                break;
-            }
+    info!("Fork detected, searching for a common ancestor between the local chain and the network chain");
```

Comment:
nit: this log message is basically the same as the comment above, you can probably delete the comment

Reply:
Fixed.